src/validateFixedIts.cpp

```diff
@@ -0,0 +1,226 @@
+
+//@HEADER
+// ***************************************************
+//
+// HPGMP: High Performance Generalized minimal residual
+//        - Mixed-Precision
+//
+// Contact:
+// Ichitaro Yamazaki         (iyamaza@sandia.gov)
+// Sivasankaran Rajamanickam (srajama@sandia.gov)
+// Piotr Luszczek            (luszczek@eecs.utk.edu)
+// Jack Dongarra             (dongarra@eecs.utk.edu)
+//
+// ***************************************************
+//@HEADER
+
+/*!
+ @file TestGMRES.cpp
+
+ HPGMP routine
+ */
+
+// Changelog
+// None
+/////////////////////////////////////////////////////////////////////////
+
+#include <fstream>
+#include <iostream>
+#include <vector>
+#include <stdexcept>
+#include "hpgmp.hpp"
+
+#include "SetupProblem.hpp"
+#include "GMRES.hpp"
+#include "GMRES_IR.hpp"
+
+#include "ValidGMRES.hpp"
+#include "mytimer.hpp"
+
+/*!
+  Test the correctness of the optimized GMRES implementation
+
+  @param[in]      argc      the "argc" parameter passed to the main() function
+  @param[in]      argv      the "argv" parameter passed to the main() function
+  @param[in]      comm      the communicator used to run validation
+  @param[inout]   test_data the data structure with the results of the test including pass/fail information
+
+  @return Returns zero on success and a non-zero value otherwise.
+
+  @see GMRES()
+ */
+
+
+template<class TestGMRESSData_type, class scalar_type, class scalar_type2, class project_type>
+int ValidGMRESFixed(const int argc, char **argv, const validation_t validation_type, comm_type comm,
+                    DeviceCtx *const dctx,
+                    const int numberOfMgLevels, const bool verbose, TestGMRESSData_type& test_data)
+{
+  if (validation_type == validation_t::standard) {
+      std::printf("Incorrect validation type!\n"); fflush(stdout);
+      return -1;
+  }
+  typedef Vector<scalar_type> Vector_type;
+  typedef SparseMatrix<scalar_type> SparseMatrix_type;
+  typedef GMRESData<scalar_type> GMRESData_type;
+
+  typedef Vector<scalar_type2> Vector_type2;
+  typedef SparseMatrix<scalar_type2> SparseMatrix_type2;
+  typedef GMRESData<scalar_type2,  project_type> GMRESData_type2;
+
+  double total_validation_time = mytimer();
+
+  //////////////////////////////////////////////////////////
+  // Setup problem
+  Geometry * geom = new Geometry;
+
+  SparseMatrix_type A;
+  GMRESData_type data;
+
+  SparseMatrix_type2 A_lo;
+  GMRESData_type2 data_lo;
+
+  Vector_type b, x;
+  SetupProblem("valid_", argc, argv, comm, dctx, numberOfMgLevels, verbose, geom, A, data, A_lo, data_lo, b, x, test_data);
+
+  //////////////////////////////////////////////////////////
+  // Solver Parameters
+  const int max_iters = validation_type == validation_t::fullscale_residual ? 1000000 : 10000;
+  const int restart_length = test_data.restart_length;
+  // If we're converging the global solver to a specific tolerance,
+  //   increase it a bit for the sake of time.
```

Comment:
This comment only applies to the fixed tolerance; not the the fixed number of iterations. It would be good to document the logic for the fixed number of iterations.

I wouldn't use the syntactic sugar `if` in this case. It will be clearer to spell out that the other option is `validation_t::fullscale_its` and group together the appropriate values of `max_iters` and `tolerance`.